test/server/config_validation/xds_fuzz.cc

```diff
@@ -0,0 +1,215 @@
+ */
+absl::optional<std::string> XdsFuzzTest::removeRoute(uint32_t route_num) {
+  std::string match = absl::StrCat("route_config_", route_num % num_routes_);
+  for (auto it = routes_.begin(); it != routes_.end(); ++it) {
+    if (it->name() == match) {
+      routes_.erase(it);
+      return match;
+    }
+  }
+  return {};
+}
+
+/**
+ * run the sequence of actions defined in the fuzzed protobuf
+ */
+void XdsFuzzTest::replay() {
+  initialize();
+
+  // set up cluster
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Cluster, "", {}, {}, {}, true));
+  sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(Config::TypeUrl::get().Cluster,
+                                                             {buildCluster("cluster_0")},
+                                                             {buildCluster("cluster_0")}, {}, "0");
+  EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().ClusterLoadAssignment, "",
+                                      {"cluster_0"}, {"cluster_0"}, {}));
+  sendDiscoveryResponse<envoy::config::endpoint::v3::ClusterLoadAssignment>(
+      Config::TypeUrl::get().ClusterLoadAssignment, {buildClusterLoadAssignment("cluster_0")},
+      {buildClusterLoadAssignment("cluster_0")}, {}, "0");
+
+  for (const auto& action : actions_) {
+    switch (action.action_selector_case()) {
+    case test::server::config_validation::Action::kAddListener: {
+      uint32_t listener_num = action.add_listener().listener_num();
+      removeListener(listener_num);
+      auto listener = buildListener(listener_num, action.add_listener().route_num());
+      listeners_.push_back(listener);
+
+      updateListener(listeners_, {listener}, {});
+      // TODO(samflattery): compareDiscoveryResponse to check ACK/NACK?
+      break;
+    }
+    case test::server::config_validation::Action::kRemoveListener: {
+      auto removed = removeListener(action.remove_listener().listener_num());
+
+      if (removed) {
+        updateListener(listeners_, {}, {*removed});
+      }
+
+      break;
+    }
+    case test::server::config_validation::Action::kAddRoute: {
+      uint32_t route_num = action.add_route().route_num();
+      auto removed = removeRoute(route_num);
+      auto route = buildRouteConfig(route_num);
+      routes_.push_back(route);
+      if (!removed) {
+        // if the route was already in routes_, don't send another request
+        updateRoute(routes_, {route}, {});
+      } else {
+        updateRoute(routes_, {}, {});
+      }
+      break;
+    }
+    case test::server::config_validation::Action::kRemoveRoute: {
+      if (sotw_or_delta_ == Grpc::SotwOrDelta::Sotw) {
+        // routes cannot be removed in SOTW updates
+        break;
+      }
+
+      auto removed = removeRoute(action.remove_route().route_num());
+      if (removed) {
+        updateRoute(routes_, {}, {*removed});
+      }
+      break;
+    }
+    default:
+      break;
+    }
+    // TODO(samflattery): makeSingleRequest here?
```

Comment:
Yeah, I think you should only bump version when you have a successful ack sequence.

Reply:
done